feat(secretstores): Add systemd-credentials plugin

srebhan opened the pull request:
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

The PR adds a secret-store plugin allowing to access systemd credentials from Telegraf (**requires systemd 254+**). This is helpful when starting Telegraf as a service without manual interaction. There the new plugin can serve as a initial secret-store to unlock others or to provide the secrets itself.

Please checkout https://systemd.io/CREDENTIALS/ for details.

Tested on ArchLinux and Ubuntu Server 23.04.

jcpunk commented:
I have some limited interest in this.

jdstrand commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,248 @@
++subsystem
+```
+
+The output should look similar to the above.
````
I suggest adding:
"The output should look similar to the above. If it says "yes" to TPM support being available then credentials can also be tied to the device by utilizing TPM sealing. See the `systemd-creds` man page for details."

jdstrand reviewed (approved):
The code and process looks fine. I've approved but added a number of comments to improve the documentation, but feel free to incorporate them and commit with another review from me.

jdstrand commented:
I forgot to mention, `systemd` 250 is pretty new. I looked at various Enterprise distros and found:
* Ubuntu 22.04 LTS has 249 (22.10 (non-LTS) has 252 so next Ubuntu LTS (24.04) will)
* Debian Bookworm has 252
* RHEL 9.2 has 252
* SUSE Enterprise Server 15 SP5 has 249
* Amazon Linux 2023 has 252
* Oracle Linux 9 has 252

I then used `lxd` to create a bookworm VM:
```
$ lxc launch images:debian/bookworm/amd64 debian-bookworm --vm -c limits.cpu=2 -c limits.memory=4GiB --console=vga
$ lxc exec debian-bookworm --user=0 --group=0 -- mkdir -m 0700 /root/.ssh # may already exist
$ lxc file push $HOME/.ssh/id_rsa.pub debian-bookworm/root/.ssh/authorized_keys --mode=0600
$ lxc exec debian-bookworm --user=0 --group=0 -- chown root:root /root/.ssh/authorized_keys
$ lxc shell debian-bookworm
$ sudo apt-get update && sudo apt-get install openssh-server
# now can ssh with 'ssh root@debian-bookworm.lxd'
```

I could then add a remove a virtual tpm device with:
```
$ lxc config device add debian-bookworm vtpm tpm path=/dev/tpm0   # shutdown the vm before doing this
$ lxc config device remove debian-bookworm vtpm                   # shutdown the vm before doing this
```

This got me to see:
```
$ sudo systemd-creds has-tpm2
yes
+firmware
+driver
+system
+subsystem
```

which was cool, but it appears I needed to do more to actually use it:
```
$ echo foo | sudo systemd-creds encrypt - /etc/telegraf/credentials/testme
Credential secret file '/var/lib/systemd/credential.secret' is not located on encrypted media, using anyway.
TPM2 support not installed: Operation not supported
TPM2 sealing didn't work, continuing without TPM2: Operation not supported
```

but I was able to solve that with:
```
$ sudo apt-get install tpm2-tools
$ echo foo | sudo systemd-creds encrypt - /etc/telegraf/credentials/testme
Credential secret file '/var/lib/systemd/credential.secret' is not located on encrypted media, using anyway.
$ sudo systemd-creds decrypt /etc/telegraf/credentials/testme
foo
```

If I shutdown the VM, remove the tpm, start it and try to access the secret, I see:
```
$ sudo systemd-creds has-tpm2
partial
-firmware
-driver
+system
+subsystem
$ sudo systemd-creds decrypt /etc/telegraf/credentials/testme
ERROR:tcti:src/tss2-tcti/tcti-device.c:452:Tss2_Tcti_Device_Init() Failed to open specified TCTI device file /dev/tpmrm0: No such file or directory 
Failed to initialize TCTI context: tcti:IO failure
```

Cool!

jdstrand commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,248 @@
+```text
+[Service]
+LoadCredentialEncrypted=http_user:/etc/telegraf/credentials/http_user
+LoadCredentialEncrypted=http_passwd:/etc/telegraf/credentials/http_passwd
````
We should consider documenting the use of `PrivateMounts=` as described in https://systemd.io/CREDENTIALS/:

"It is a good idea to also enable mount namespacing for services that process credentials configured this way. If so, the runtime credential directory of the specific service is not visible to any other service. Use PrivateMounts= as minimal option to enable such namespacing. Note that many other sandboxing settings (e.g. ProtectSystem=, ReadOnlyPaths= and similar) imply PrivateMounts=, hence oftentimes it’s not necessary to set this option explicitly."

jdstrand replied:
This *should* be safe since the system mounts should be visible to telegraf but new mounts (like the one for the decrypted files) are not propagated outside of telegraf's private mount namespace, but will need testing. From https://manpages.ubuntu.com/manpages/lunar/en/man5/systemd.exec.5.html
```
PrivateMounts=
           Takes a boolean parameter. If set, the processes of this unit will be run in their own
           private file system (mount) namespace with all mount propagation from the processes
           towards the host's main file system namespace turned off. This means any file system
           mount points established or removed by the unit's processes will be private to them
           and not be visible to the host. However, file system mount points established or
           removed on the host will be propagated to the unit's processes. See
           [mount_namespaces](https://manpages.ubuntu.com/manpages/lunar/en/man7/mount_namespaces.7.html)(7) for details on file system namespaces. Defaults to off.
```

srebhan replied:
I don't think we can do this as it also means that Telegraf cannot read files in the file-system which is required by various plugins...

jdstrand replied:
That's the thing; my interpretation of the man page is that the system's private mounts all propagate into the container, but mounts within the container do not propagate into the host. This is also in the man page:
```
           When turned on, this executes three operations for each invoked process: a new
           CLONE_NEWNS namespace is created, after which all existing mounts are remounted to
           MS_SLAVE to disable propagation from the unit's processes to the host (but leaving
           propagation in the opposite direction in effect). Finally, the mounts are remounted
           again to the propagation mode configured with MountFlags=, see below.
```


This is why the flag is meaningful for this feature: without it, the mount is available to others but with it, it is not available to other services (that use PrivateMounts).

I did look into this further and found that:
* when the service is running and `LoadCredentialsEncrypted=...` is used, `$CREDENTIALS_DIRECTORY` is `/run/credentials/telegraf.service` in the service, and this directory has 0500 permissions with `telegraf` as the owner (decrypted files are 0500 with `telegraf` as owner)
* when the service is running and `LoadCredentialsEncrypted=...` is used, the decrypted credentials are in the root-protected `/run/credentials/telegraf.service` directory on the host
* when the service is running and `LoadCredentialsEncrypted=...` is used and `PrivateMounts` IS NOT set, everything in `/run/credentials` on the host is available in `/run/credentials` in the service
* when the service is running and `LoadCredentialsEncrypted=...` is used and `PrivateMounts` IS set, only  `/run/credentials/telegraf.service` on the host is available in `/run/credentials` in the service
* when services that specify `LoadCredentialsEncrypted=...` aren't running, then the decrypted credentials for those services are NOT present in `/run/credentials` on the host (or any running services)

Importantly, this has a few implications:
1. root on the host will always be able to see all the credentials via `/run/credentials` (but also by entering the running the `systemd-creds` command, entering mount namespaces, adjusting unit files, etc). There is no protection against unrestricted root processes on the host
2. with or without `PrivateMounts` being set, the decrypted credentials have DAC permissions set to the `User=` in the systemd unit. Specifically for telegraf, only processes as the `telegraf` user can see the decrypted credentials in `/run/credentials/telegraf.service` on the host or within the service. Other non-root processes cannot see the credentials
3. with `PrivateMounts` NOT SET in a service's unit file that runs as ROOT, the root user can see all currently decrypted credentials for running services (eg, the `foo.service` doesn't specify `User=foo` (or specifies `User=root`), this service can see `/run/credentials/foo.service/*`, `/run/credentials/telegraf.service/*` and any other decrypted credentials for running services)
4. with `PrivateMounts` NOT SET in a service's unit file that runs as NON-ROOT, this non-root user can see all currently decrypted credentials for running services that are accessible to this non-root user
5. with `PrivateMounts` SET in a service's unit file that runs as ROOT, the root user can only see its own decrypted credentials (eg, the `foo.service` doesn't specify `User=foo` (or specifies `User=root`), this service can see `/run/credentials/foo.service/*` but not `/run/credentials/telegraf.service/*` or any other decrypted credentials for running services)
6. with `PrivateMounts` SET in a service's unit file that runs as NON-ROOT, this service can only see `/run/credentials/foo.service/*` but not `/run/credentials/telegraf.service/*` or any other decrypted credentials for running services regardless of if they run as the same user or not)

Practically speaking, this feature does not protect against attacker-controlled root processes or root running services that don't set `PrivateMounts` as these processes can always see the decrypted credentials. `PrivateMounts` adds some protection to the telegraf service against attacker-controlled non-root running services. It adds considerable protection against attacker-controlled root running services when those root running services also specify `PrivateMounts` in their unit files.

srebhan commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,248 @@
+After setting up the encryption key we can create a new credential using
+
+```shell
+# echo -n "john-doe-jr" | systemd-creds encrypt - /etc/telegraf/credentials/http_user
````
I thought about it and came to the conclusion that it is easier to give a single dir in the service file instead of adding _multiple_ entries containing encrypted credentials. Both is possible and supported but I want to keep the examples clean here...

srebhan commented on plugins/secretstores/systemd/README.md:
```diff
@@ -0,0 +1,248 @@
+credential/secret's name rule.
+
+We can now add more secrets. To avoid potentially leaking the plain-text
+credentials through command-history or showing it on the screen we use
```
Right. I didn't want to give a false sense of security by including this in the docs...

srebhan commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,248 @@
+[Service]
+LoadCredentialEncrypted=http_user:/etc/telegraf/credentials/http_user
+LoadCredentialEncrypted=http_passwd:/etc/telegraf/credentials/http_passwd
+```
````
That's really strange. For me (systemd 254 on ArchLinux) this works without creating the dir... Will add a note...

srebhan commented on plugins/secretstores/systemd/systemd.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (s *Systemd) Set(_, _ string) error {
+	return errors.New("secret-store does not support creating secrets")
```
This will automatically added by Telegraf's logging facilities...

srebhan commented on plugins/secretstores/systemd/systemd.go:
```diff
@@ -0,0 +1,136 @@
+	}
+	s.Log.Debugf("Found systemd version %d...", version)
+	if version < systemdMinimumVersion {
+		return fmt.Errorf("systemd version %d below minimum version %d", version, systemdMinimumVersion)
```
See above. That should be added by Telegraf's logging...

jdstrand reviewed (changes requested):
Based on my investigation of `PrivateMounts`, I'm now changing this to 'Request changes' so the documentation can be updated to recommend using `PrivateMounts`. Feel free to adjust the unit file in the deb and rpm packaging to specify it by default (but that is not a blocker for this PR and can come at a later time). Please also mention in the docs why it is recommended (and ideally that it can offer protections when other units also specify it).

srebhan commented on plugins/secretstores/systemd/README.md:
```diff
@@ -0,0 +1,248 @@
+# Systemd Secret-Store Plugin
+
+The `systemd` plugin allows to utilize credentials and secrets provided by
+[systemd][] to the Telegraf service. Systemd ensures that only the intended
```
No you don't need anything in the second brackets as this is a named link and the link-name is what is in the first brackets. Tested and works.

srebhan commented:
@jdstrand thanks for the review. I hope I addressed all your points.
@powersj I added `PrivateMounts = true` to the service as requested by Jamie and I agree that this should (from documentation) work. If we see issues with not being able to access files we likely need to revert that. Just as a heads-up.

jdstrand commented on plugins/secretstores/systemd/README.md:
```diff
@@ -0,0 +1,263 @@
+`systemd-creds encrypt`) does not match the one used in the
+`LoadCredentialEncrypted` statement.
+
+In case you are having trouble to reference credentials in Telegraf, you should
```
s/to reference/referencing/

jdstrand reviewed (approved):
Thanks for the updates! I've given my approval so you're not blocked on me, but I gave some more suggestions to the readme.

srebhan commented:
@jdstrand thanks for the feedback. I used all your suggestions so you might or might not want to take another look and approve!?

jdstrand reviewed (approved):
Thanks for the changes!

powersj commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,268 @@
+```
+
+to your Telegraf configuration. Assuming the two example credentials
+`http_user` and `http_password` you can now use those as secrets via
````
```suggestion
`http_user` and `http_password` you can now use those as secrets via
the following config:
```

powersj commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,268 @@
+
+```
+
+in your plugins.
+
````
```suggestion
```

powersj commented on plugins/secretstores/systemd/systemd.go:
```diff
@@ -0,0 +1,136 @@
+	if err != nil {
+		return fmt.Errorf("unable to detect systemd version: %w", err)
+	}
+	s.Log.Debugf("Found systemd version %d...", version)
```
```suggestion
	s.Log.Debugf("Found systemd version %d", version)
```

powersj commented on plugins/secretstores/systemd/systemd.go:
```diff
@@ -0,0 +1,136 @@
+
+	// By default the credentials directory is passed in by systemd
+	// via the "CREDENTIALS_DIRECTORY" environment variable.
+	defaultPath := os.Getenv("CREDENTIALS_DIRECTORY")
+	if defaultPath == "" {
+		s.Log.Warn("CREDENTIALS_DIRECTORY environment variable undefined. Make sure credentials are setup correctly!")
+		if s.Path == "" {
+			return errors.New("'path' required without CREDENTIALS_DIRECTORY")
+		}
+	}
```
The warning message is throwing me off. If a user sets the path correctly, do they care about the CREDENTIALS_DIRECTORY variable?

srebhan replied:
This is a "heads-up" warning. Under normal circumstances, systemd sets the `CREDENTIALS_DIRECTORY` in any case. So you should never end-up there if stuff is working correctly and credentials are initialized with systemd...

powersj reviewed (commented):
Thank you for the work on this and sorry this took so long. Mainly questions for my education and to verify we are on the same page.

srebhan commented on scripts/telegraf.service:
```diff
@@ -14,6 +14,8 @@ Restart=on-failure
 RestartForceExitStatus=SIGPIPE
 KillMode=control-group
 LimitMEMLOCK=8M:8M
+PrivateMounts=true
```
From the [man-page](https://www.freedesktop.org/software/systemd/man/systemd.unit.html):
> Unit files may contain additional options on top of those listed here. If systemd encounters an unknown option, it will write a warning log message but continue loading the unit. 

Should I comment the option the same way as `LoadCredentialEncrypted`?

powersj replied:
Yeah - can we comment out the option and make this opt-in please?

jdstrand replied:
Well, `PrivateMounts` is an important security feature. I noticed that grafana's systemd unit packaging has directives that for newer `systemd` and when using the unit on an old `systemd`, `systemd` complains in the logs, but otherwise ignores it. Is this an option (I've not tested the behavior with `PrivateMounts`)?

powersj replied:
I have an absolute fear of regressing or changing the package behavior. As both have you have now mentioned it should just ignore them, so maybe I am being a little too aggressive with avoiding the change if that is the case. Happy to keep it then.

srebhan commented on scripts/deb/post-install.sh:
```diff
@@ -42,6 +42,11 @@ if [[ ! -d /etc/telegraf/telegraf.d ]]; then
 
+# Add credentials directory
+if [[ ! -d /etc/telegraf/credentials ]]; then
+    mkdir -p /etc/telegraf/credentials
```
Lennart suggested to place the credentials in `/etc/credstore.encrypted` and prefix them with `telegraf.` e.g. `telegraf.masterpassword` what do you think?

powersj replied:
@srebhan did you want to make this change?

srebhan replied:
Not needed anymore.

srebhan commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,268 @@
+To actually provide credentials to the Telegraf service, you need to list them
+in the service file. You can use
+
+```shell
+$ sudo systemctl edit telegraf
+...
+```
+
+to overwrite parts of the service file. On some systems you need to create the
+overriding directory `/etc/systemd/system/telegraf.service.d` first. The
+resulting override can be found in
+`/etc/systemd/system/telegraf.service.d/override.conf`. The following is an
+example for the content of the file
````
That is what `systemctl edit` does... At least on my system... :-)

powersj replied:
Right, I saw the same, so I was curious when/why someone would need to create the directory first.

jdstrand replied:
This was added because in my testing on a Debian system, the directory wasn't created.

powersj replied:
Ah ok, interesting!

Hipska commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,267 @@
+
+Now setup the credentials by creating the root key.
+
+```shell
````
```suggestion
```text
```

Hipska commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,267 @@
+
+After setting up the encryption key we can create a new credential using
+
+```shell
````
```suggestion
```text
```

(it would haven bee fine if the output from the command wasn't included..)

Hipska commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,267 @@
+We can now add more secrets. To avoid potentially leaking the plain-text
+credentials through command-history or showing it on the screen we use
+
+```shell
````
```suggestion
```text
```

Hipska commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,267 @@
+
+```shell
+$ sudo systemctl edit telegraf
+...
````
```suggestion
```

Hipska commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,267 @@
+`/etc/systemd/system/telegraf.service.d/override.conf`. The following is an
+example for the content of the file
+
+```text
````
```suggestion
```ini
```

Hipska commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,267 @@
+In case you are having trouble referencing credentials in Telegraf, you should
+check what is available via
+
+```shell
````
```suggestion
```text
```

Hipska commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,267 @@
+
+To get the actual value of a credential use
+
+```shell
````
```suggestion
```text
```

Hipska commented on plugins/secretstores/systemd/systemd_test.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+func TestEmptyPath(t *testing.T) {
+	getSystemdVersion = func() (int, error) { return systemdMinimumVersion, nil }
```
Maybe extract definition of this function instead of redefining in each test?

powersj commented on plugins/secretstores/systemd/README.md:
````diff
@@ -0,0 +1,267 @@
+for protecting the credentials, you should first make sure that it is
+available using
+
+```shell
````
While not a script it is showing a command inside a shell. By using shell we can get syntax highlighting.

Hipska replied:
It is also showing the output, then it is not a shell script anymore. If you want highlights of the command, then the output should be a separate code block
